test/common/config/new_grpc_mux_impl_test.cc

```diff
@@ -179,6 +179,90 @@ TEST_F(NewGrpcMuxImplTest, ReconnectionResetsNonceAndAcks) {
+// Regression test of https://github.com/envoyproxy/envoy/issues/16063.
+TEST_F(NewGrpcMuxImplTest, ReconnectionResetsWildcardSubscription) {
+  Event::MockTimer* grpc_stream_retry_timer{new Event::MockTimer()};
+  Event::MockTimer* ttl_mgr_timer{new NiceMock<Event::MockTimer>()};
+  Event::TimerCb grpc_stream_retry_timer_cb;
+  EXPECT_CALL(dispatcher_, createTimer_(_))
+      .WillOnce(
+          testing::DoAll(SaveArg<0>(&grpc_stream_retry_timer_cb), Return(grpc_stream_retry_timer)))
+      // Happens when adding a type url watch.
+      .WillRepeatedly(Return(ttl_mgr_timer));
+  setup();
+  InSequence s;
+  const std::string& type_url = Config::TypeUrl::get().ClusterLoadAssignment;
+  auto foo_sub = grpc_mux_->addWatch(type_url, {}, callbacks_, resource_decoder_, {});
+  EXPECT_CALL(*async_client_, startRaw(_, _, _, _)).WillOnce(Return(&async_stream_));
+  // Send a wildcard request on new connection.
+  expectSendMessage(type_url, {}, {});
+  grpc_mux_->start();
+
+  // An helper function to create a response with a single load_assignment resource
+  // (load_assignment's cluster_name will be updated).
+  envoy::config::endpoint::v3::ClusterLoadAssignment load_assignment;
+  auto create_response = [&load_assignment, &type_url](const std::string& name,
+                                                       const std::string& version,
+                                                       const std::string& nonce)
+      -> std::unique_ptr<envoy::service::discovery::v3::DeltaDiscoveryResponse> {
+    auto response = std::make_unique<envoy::service::discovery::v3::DeltaDiscoveryResponse>();
+    response->set_type_url(type_url);
+    response->set_system_version_info(version);
+    response->set_nonce(nonce);
+    auto res = response->add_resources();
+    res->set_name(name);
+    res->set_version(version);
+    load_assignment.set_cluster_name(name);
+    res->mutable_resource()->PackFrom(load_assignment);
+    return response;
+  };
+
+  // Send a response with a single resource that should be received by Envoy,
+  // followed by an ack with the nonce.
+  {
+    auto response = create_response("x", "1000", "111");
+    EXPECT_CALL(callbacks_, onConfigUpdate(_, _, "1000"))
+        .WillOnce(Invoke([&load_assignment](const std::vector<DecodedResourceRef>& added_resources,
+                                            const Protobuf::RepeatedPtrField<std::string>&,
+                                            const std::string&) {
+          EXPECT_EQ(1, added_resources.size());
+          EXPECT_TRUE(
+              TestUtility::protoEqual(added_resources[0].get().resource(), load_assignment));
+        }));
+    // Expect a reply with the nonce.
+    expectSendMessage(type_url, {}, {}, "111");
+    grpc_mux_->onDiscoveryResponse(std::move(response), control_plane_stats_);
+  }
+  // Send another response with a different resource, but where EDS is paused.
+  auto resume_cds = grpc_mux_->pause(type_url);
```

Comment:
Why did you add an update with a paused mux to this test?

Reply:
This is done to emulate a disconnect before an ack is sent from the client, and ensure that the message on reconnect does include the new resource.
I'll add some comments to this test.